File: neptun_alarm_app/test/radar_repository_test.dart
import 'dart:convert';

import 'package:flutter_test/flutter_test.dart';
import 'package:neptun_alarm_app/features/radar/data/radar_repository.dart';

void main() {
  group('RadarRepository.parseThreatsEnvelope', () {
    test('top-level JSON array', () {
      final out = RadarRepository.parseThreatsEnvelope([
        {'id': 'a', 'threatType': 'shahed'},
      ]);
      expect(out.length, 1);
      expect(out.first['threatType'], 'shahed');
    });

    test('map with threats key', () {
      final raw = jsonDecode(
        '{"threats":[{"id":"1","type":"raketa"}]}',
      );
      final out = RadarRepository.parseThreatsEnvelope(raw);
      expect(out.length, 1);
      expect(out.first['type'], 'raketa');
    });

    test('map with markers key', () {
      final raw = {'markers': <Map<String, Object>>[]};
      expect(RadarRepository.parseThreatsEnvelope(raw), isEmpty);
    });

    test('unsupported shape returns empty', () {
      expect(RadarRepository.parseThreatsEnvelope('x'), isEmpty);
    });
  });

  group('RadarRepository.normalizeThreatMarker', () {
    test('maps V10 snake_case fields', () {
      final out = RadarRepository.normalizeThreatMarker({
        'id': 't1',
        'trackQualityScore': 0.82,
        'formationId': 'wave-7',
        'maneuverDetected': true,
        'predictedImpact': {'eta_minutes': 12},
      });
      expect(out['track_quality_score'], 0.82);
      expect(out['formation_id'], 'wave-7');
      expect(out['maneuver_detected'], isTrue);
      expect(out['predicted_impact'], isA<Map>());
    });
  });

  group('RadarRepository.countActiveRegionsFromAlarmStream', () {
    test('counts rows with non-empty activeAlerts', () {
      final n = RadarRepository.countActiveRegionsFromAlarmStream([
        {'activeAlerts': ['a']},
        {'activeAlerts': []},
        {'region': 'x'},
      ]);
      expect(n, 1);
    });
  });
}

File: neptun_expo_app/src/features/radar/data/radarRepository.test.ts
import {
  countActiveRegionsFromAlarmStream,
  normalizeThreatMarker,
  parseThreatsEnvelope,
} from './radarParsing';

function assert(cond: boolean, msg: string): void {
  if (!cond) throw new Error(msg);
}

// Flutter `radar_repository_test.dart` parity
assert(parseThreatsEnvelope([{ id: 'a', threatType: 'shahed' }]).length === 1, 'array envelope');
assert(
  parseThreatsEnvelope({ threats: [{ id: '1', type: 'raketa' }] }).length === 1,
  'threats key',
);
assert(parseThreatsEnvelope({ markers: [] }).length === 0, 'empty markers');
assert(parseThreatsEnvelope('x').length === 0, 'unsupported');

const norm = normalizeThreatMarker({
  id: 't1',
  trackQualityScore: 0.82,
  formationId: 'wave-7',
  maneuverDetected: true,
  predictedImpact: { eta_minutes: 12 },
});
assert(norm.track_quality_score === 0.82, 'quality');
assert(norm.formation_id === 'wave-7', 'formation');
assert(norm.maneuver_detected === true, 'maneuver');
assert(typeof norm.predicted_impact === 'object', 'impact');

assert(
  countActiveRegionsFromAlarmStream([
    { activeAlerts: ['a'] },
    { activeAlerts: [] },
    { region: 'x' },
  ]) === 1,
  'alarm count',
);

console.log('radarRepository.test.ts: ok');
